src/main/kotlin/org/rust/toml/CargoTomlDependenciesCompletionProvider.kt

```diff
@@ -0,0 +1,136 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.toml
+
+import com.intellij.codeInsight.completion.*
+import com.intellij.codeInsight.lookup.LookupElement
+import com.intellij.codeInsight.lookup.LookupElementBuilder
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.tree.IElementType
+import com.intellij.psi.util.PsiTreeUtil
+import com.intellij.util.ProcessingContext
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.elementType
+import org.rust.lang.core.psi.ext.isAncestorOf
+import org.toml.lang.psi.*
+
+abstract class AbstractTomlKeyValueCompletionProvider : CompletionProvider<CompletionParameters>() {
+    override fun addCompletions(parameters: CompletionParameters, context: ProcessingContext?, result: CompletionResultSet) {
+        val parent = parameters.position.parent ?: return
+        if (parent is TomlKey) {
+            completeDependencyKey(parent, result)
+        } else {
+            val keyValue = parent.ancestorOrSelf<TomlKeyValue>() ?: return
+            // If a value is already present we should ensure that the value is a literal
+            // and the caret is inside the value to forbid completion in cases like
+            // `key = "" <caret>`
+            val value = keyValue.value
+            if (value != null && (value !is TomlLiteral || !value.isAncestorOf(parameters.position))) return
+
+            completeDependencyValue(keyValue, result)
+        }
+    }
+
+    protected abstract fun completeDependencyKey(key: TomlKey, result: CompletionResultSet)
+
+    protected abstract fun completeDependencyValue(keyValue: TomlKeyValue, result: CompletionResultSet)
+}
+
+/** @see CargoTomlPsiPattern.inDependencyKeyValue */
+class CargoTomlDependenciesCompletionProvider : AbstractTomlKeyValueCompletionProvider() {
```

Comment:
Let's use the same naming style for these providers and replace `Dependencies` to `Dependency`